### PR TITLE
feat: add oauth2 token exchange

### PR DESCRIPTION
**Pull request by code-asher**
This adds a new identity provider package and hooks it into coderd routes along with some test refactoring.  See commit messages for more details.

There is more work to be done (tracking in https://github.com/coder/coder/pull/11609) but I think this is fully featured enough to be merged in.  All the routes are only accessible in dev mode.

It is getting pretty big, although 60% is HTML, tests, and generated code!

Blocked by:
- https://github.com/coder/coder/pull/12196

Blocks:
- https://github.com/coder/coder/pull/11780
- https://github.com/coder/coder/pull/12127

### CODE REVIEW

File: enterprise/coderd/identityprovider/tokens.go
Comment (code-asher) [inline enterprise/coderd/identityprovider/tokens.go:244]: I am using the app ID for this but I think the salt is supposed to be unique per secret.  I could prepend the ID of every secret to the secret, although that makes them pretty long.  Only the client secret is ever seen by a person though.

File: enterprise/coderd/identityprovider/tokens.go
Comment (Emyrk) [inline enterprise/coderd/identityprovider/tokens.go:122]: Ooooh I now see why you did this. My above comment about using `userpassword.Hash` remains, I think we should just prefix this stuff with some id and use `userpassword.Compare(hashedSecret, secret)` to validate.

This would mean you would have to store an extra id in the database...

So currently a secret is 40 random characters. We should prefix it with something.

Before: `EA9ILdVv0wVgXo840dFZV4UX21dtZsl5rziDiiU7`

New: `coder_io928hVamL_EA9ILdVv0wVgXo840dFZV4UX21dtZsl5rziDiiU7` (or something like this).

You would have to store `io928hVamL` in the db, so when we are given the secret, we split on `_`, find the `id` and do the search with that. This column also needs a `unique` constraint.

File: enterprise/coderd/identityprovider/tokens.go
Comment (Emyrk) [inline enterprise/coderd/identityprovider/tokens.go:128]: Same as above hash comment

File: enterprise/coderd/identityprovider/tokens.go
Comment (Emyrk) [inline enterprise/coderd/identityprovider/tokens.go:156]: Maybe we disable auto refresh for oauth type api keys :shrug:.

File: enterprise/coderd/identityprovider/tokens.go
Comment (Emyrk) [inline enterprise/coderd/identityprovider/tokens.go:212]: Same hash comment as above

File: enterprise/coderd/identityprovider/tokens.go
Comment (Emyrk) [inline enterprise/coderd/identityprovider/tokens.go:235]: I see why you do not have Expiry, but that usually means it never expires.

I think we should add a expiry and disable auto refreshing of these api keys.

File: enterprise/coderd/identityprovider/tokens.go
Comment (Emyrk) [reply enterprise/coderd/identityprovider/tokens.go:245]: I think we need to ditch this and use `userpassword.Hash`. I think it might be fine in practice, I just rather stick to the hash method we already know and use. I left another comment about how to handle the fact that `userpassword.Hash` is non-deterministic. 

Some notes on the security:
- `userpassword.Hash` is stronger because users might have the same password (`password123`) and these should be hashed to different values.
- This has function takes in a `secret` which already has a lot of randomness and is longer, so it does not need as strong of a hash function.

Still would rather use what we already use.

Comment (Emyrk) [reply enterprise/coderd/identityprovider/tokens.go:245]: > I am using the app ID for this but I think the salt is supposed to be unique per secret. 

For the best security yes. You could use `sha256(appID + userID)[:10]` or something deterministic but more unique. The best is a random salt, like done in `userpassword.Hash()`

> I could prepend the ID of every secret to the secret, although that makes them pretty long

We can make a new ID that is shorter. I put this comment above, but it just came to me if you want to use the `uuid` already and make it smaller, just do: `sha256(id)[:10]` and encode it with like base64 or something. The downside to just making a new ID for this purpose those is that base64 has is a non deterministic length encoding for any given input. (well you can predict it, but it is a pain).

So if we just choose random characters, then the ID is always a fixed length. We should not just use a `uuid` string though, as it is a terribly inefficient use of characters and is way to long.

File: enterprise/coderd/oauth2.go
Comment (Emyrk) [inline enterprise/coderd/oauth2.go:90]: We might update the db app, and we should pass the whole db app to the db2sdk func always. This is more resistance to migrations on the BE.


```suggestion
	var sdkApps []codersdk.OAuth2ProviderApp
	for _, app := range userApps {
		sdkApps = append(sdkApps, db2sdk.OAuth2ProviderApp(api.AccessURL, app.OAuth2ProviderApp))
	}
	httpapi.Write(ctx, rw, http.StatusOK, sdkApps)
```

File: enterprise/coderd/oauth2.go
Comment (Emyrk) [inline enterprise/coderd/oauth2.go:244]: hash comments above

**Review by Emyrk — CHANGES_REQUESTED** [review]
Consolidating and adding some hash function notes.

We should use `userpasword.Hash()`

The user secret is currently 40 random characters, let's spruce that up.
- Add a `coder_` prefix so a random key in a text file is more easily recognized. 
- Add an `id` section of 10(?) random characters with a unique constraint.
- Optionally... if you want to be really cool, add a 2 character checksum at the end, so if someone hand types this we can detect typos :wink:

File: coderd/httpapi/queryparams.go
Comment (code-asher) [reply coderd/httpapi/queryparams.go:249]: Yeah I had mixed feelings about this; it felt surprising to me that I could write `Required("client_id")` and it could come through empty.  My thinking was that if I want to require a query param I also want a value.

I did check existing usages; it is used for `reconnect` on the terminal (which is invalid if blank) and for `start_time` and `end_time` on an insights endpoint which also cannot be blank (it will fail to be parsed into a time).

But at the same time I see the distinction between required and empty.  An empty query value is probably only useful as a boolean where the presence of the key means `true`, and I could see requiring a `true` for something like `?acceptTerms` or something like that before you can use the endpoint.

So let me rename this to `RequiredNotEmpty`, that will leave room for `Required` if we do need that some day.

File: codersdk/oauth2.go
Comment (code-asher) [reply codersdk/oauth2.go:219]: Great point, I will add a note to implement that.

Yup, it is 1 token per user per app.

File: enterprise/coderd/identityprovider/authorize.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/authorize.go:58]: Ooo good point, let me move this.

File: enterprise/coderd/identityprovider/authorize.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/authorize.go:61]: Makes sense!  That does seem better.

Comment (code-asher) [reply enterprise/coderd/identityprovider/authorize.go:61]: Oh right I needed to check the length from the caller to tell if it was the `url.Parse(callback)` that failed (although that should not happen because we validate before storing it in the db) or if it was a validation error so I can give a different message and error code.  I think I will just lift the parse out of this func.

File: enterprise/coderd/identityprovider/authorize.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/authorize.go:69]: Gah, thanks, I keep mixing up my doc styles.

File: enterprise/coderd/identityprovider/authorize.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/authorize.go:101]: This is actually a relief lol

File: enterprise/coderd/identityprovider/authorize.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/authorize.go:110]: Oh yup you are absolutely right.  I think I got confused thinking I had to delete to cascade something but the app codes are all on their own, unlike tokens.  Will change!

File: enterprise/coderd/identityprovider/authorize.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/authorize.go:117]: Ah yup good point, will do.

File: enterprise/coderd/identityprovider/middleware.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/middleware.go:28]: Oh wow good to know.

File: enterprise/coderd/identityprovider/middleware.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/middleware.go:93]: I refactored so we are always checking `err`.  I do still have separate error since one is if the callback URL fails to parse, which should not happen but is technically a different error if it does.

File: enterprise/coderd/identityprovider/tokens.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/tokens.go:90]: Oh yeah, I did have it this way but I am validating the grant type in the parser now so it is not possible to get anything other than a valid type or empty string.

File: enterprise/coderd/identityprovider/tokens.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/tokens.go:235]: Oh yeah I was trying to see if omitting it would make Backstage stop trying to refresh, but it tries to anyway.  I wanted to merge this and work on refresh next but maybe I have to add it now.

File: enterprise/coderd/oauth2.go
Comment (code-asher) [reply enterprise/coderd/oauth2.go:90]: Ahhh makes sense!  Much better.

**Comment by code-asher** [comment]
Moved to a draft while I work on:

1. [ ] Changing the hashing
2. [ ] Adding the refresh flow

File: enterprise/coderd/identityprovider/authorize.go
Comment (Emyrk) [reply enterprise/coderd/identityprovider/authorize.go:61]: You can also return a sentinel error. So either a unique type or a unique message.

Comment (code-asher) [reply enterprise/coderd/identityprovider/authorize.go:61]: Ohh right that is smart!

File: coderd/httpapi/queryparams.go
Comment (code-asher) [reply coderd/httpapi/queryparams.go:249]: Some new ones popped up after I merged main: `template_id`, `workspace_id`, and `agent_id` on some endpoints, they all make database queries with these values so I think they would just return "no rows" or a 404 or whatever, so this change will make it return "invalid query params, template_id is required" instead which I think is fair, but it does change the semantics a little from a 404 to a 400.

**Comment by code-asher** [comment]
Rebased, but only commits from [Delete some oidc helper comments](https://github.com/coder/coder/pull/11778/commits/cbce34ac532022488576610854888dfc6c881a6a) onward are new.  They should all be self-contained enough to review individually, if that appeals.

I really like the idea of a checksum but I have not implemented it yet.  I was reading up some check digit algorithms and thinking maybe of using crc8 or something but got hit with decision paralysis so I am leaving it for the moment.  :sweat:

**Review by Emyrk — APPROVED** [review]
This is really good :+1:. 

Just 1 thing to address before I approve it in `refreshTokenGrant`. Is there a leak of refresh codes?

**Review by Emyrk — CHANGES_REQUESTED** [review]
whoops meant to do request changes

File: coderd/httpapi/queryparams.go
Comment (code-asher) [reply coderd/httpapi/queryparams.go:140]: Good idea!

File: codersdk/oauth2.go
Comment (code-asher) [reply codersdk/oauth2.go:215]: Oh really good point, let me change it now real quick.

Comment (code-asher) [reply codersdk/oauth2.go:215]: Oh right I had it here to have the post/delete paradigm on the same endpoint.  Maybe I should actually move this back to `/api/v2` since it is our own thing, not part of the whole oauth flow.

File: enterprise/coderd/identityprovider/middleware.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/middleware.go:49]: Oh yeah that does sound like a nightmare.  Hmmm...so `referer` is only required to detect the `allow` part of it, and it could be blank on the initial HTML render if opened from a desktop app or something I think.  So maybe I error if both redirected is set and referer is blank, since the redirected param indicates the user might have pressed the button.

There is a todo for cURL at the top of the function!  But I should move it down next to the HTML part.

File: codersdk/oauth2.go
Comment (Emyrk) [reply codersdk/oauth2.go:215]: This is not an oauth2 spec endpoint right? 

Eh just keep it here and comment that it is not oauth 2 spec. If you feel this has to be a static endpoint though, then just comment why it should be. I'll defer to your judgement

File: enterprise/coderd/identityprovider/middleware.go
Comment (Emyrk) [reply enterprise/coderd/identityprovider/middleware.go:49]: > So maybe I error if both redirected is set and referer is blank, since the redirected param indicates the user might have pressed the button.

This is perfect. I was thinking the exact same thing on why it would be difficult. But this solution is perfect for now imo.

We had a customer that bitbucket had this error because their proxy was nuking the Referrer header. It was the hardest thing in the world to debug.

File: enterprise/coderd/identityprovider/revoke.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/revoke.go:42]: I think it was because there is no body.

File: enterprise/coderd/identityprovider/tokens.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/tokens.go:76]: It really is!!  :heart:

File: enterprise/coderd/identityprovider/tokens.go
Comment (code-asher) [reply enterprise/coderd/identityprovider/tokens.go:60]: Yeah you are right, let me add that.

File: enterprise/coderd/identityprovider/tokens.go
Comment (Emyrk) [reply enterprise/coderd/identityprovider/tokens.go:366]: Ah! If it cascades, we are good.

**Review by Emyrk — APPROVED** [review]

**Pull request merged**
